File: src-tauri/src/shared/schema.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

fn is_valid_selection_type(st: &str) -> bool {
    st == "single_with_input" || st == "multiple_with_input"
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AskQuestion {
    pub question: String,
    pub options: Vec<String>,
    #[serde(rename = "selectionType")]
    pub selection_type: String,
}

impl AskQuestion {
    pub fn validate(&self) -> Result<(), String> {
        if !is_valid_selection_type(&self.selection_type) {
            return Err(format!(
                "Invalid selectionType '{}'. Only 'single_with_input' and 'multiple_with_input' are allowed.",
                self.selection_type
            ));
        }
        if self.options.is_empty() {
            return Err("Ask questions require at least one option. Use 'single_with_input' or 'multiple_with_input'.".to_string());
        }
        Ok(())
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Task {
    pub id: String,
    pub text: String,
    pub status: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ServerEventChunk {
    #[serde(rename = "sessionId")]
    pub session_id: String,
    #[serde(rename = "eventId")]
    pub event_id: u64,
    pub timestamp: u64,
    #[serde(flatten)]
    pub payload: EventPayload,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type")]
pub enum EventPayload {
    #[serde(rename = "text_delta")]
    TextDelta { content: String },

    #[serde(rename = "thinking_delta")]
    ThinkingDelta { content: String },

    #[serde(rename = "thinking_end")]
    ThinkingEnd,

    #[serde(rename = "tool_start")]
    ToolStart {
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        args: Value,
    },

    #[serde(rename = "tool_call_delta")]
    ToolCallDelta {
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        #[serde(rename = "argumentsText")]
        arguments_text: String,
    },

    #[serde(rename = "tool_output")]
    ToolOutput {
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "logLine")]
        log_line: String,
    },

    #[serde(rename = "tool_result")]
    ToolResult {
        #[serde(rename = "toolId")]
        tool_id: String,
        result: String,
        #[serde(rename = "isError")]
        is_error: bool,
    },

    #[serde(rename = "tool_end")]
    ToolEnd {
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "exitCode")]
        exit_code: i32,
    },

    #[serde(rename = "ask_request")]
    AskRequest {
        #[serde(rename = "toolId")]
        tool_id: String,
        title: String,
        question: String,
        options: Vec<String>,
        #[serde(rename = "selectionType")]
        selection_type: String,
        #[serde(rename = "questions", default)]
        questions: Vec<AskQuestion>,
    },

    #[serde(rename = "task_update")]
    TaskUpdate { tasks: Vec<Task> },

    #[serde(rename = "subagent_start")]
    SubagentStart {
        #[serde(rename = "parentSessionId")]
        parent_session_id: String,
        #[serde(rename = "parentToolCallId")]
        parent_tool_call_id: String,
        #[serde(rename = "subSessionId")]
        sub_session_id: String,
        title: String,
        message: String,
        #[serde(rename = "startedAt")]
        started_at: u64,
    },

    #[serde(rename = "subagent_end")]
    SubagentEnd {
        #[serde(rename = "parentSessionId")]
        parent_session_id: String,
        #[serde(rename = "parentToolCallId")]
        parent_tool_call_id: String,
        #[serde(rename = "subSessionId")]
        sub_session_id: String,
        result: String,
        #[serde(rename = "isError")]
        is_error: bool,
    },

    #[serde(rename = "permission_request")]
    PermissionRequest {
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        reason: String,
    },

    #[serde(rename = "runtime_status")]
    RuntimeStatus {
        state: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
        message: String,
        attempt: u32,
        #[serde(rename = "retryInSeconds", skip_serializing_if = "Option::is_none")]
        retry_in_seconds: Option<u32>,
        #[serde(rename = "retryAt", skip_serializing_if = "Option::is_none")]
        retry_at: Option<u64>,
    },

    #[serde(rename = "heartbeat")]
    Heartbeat,

    #[serde(rename = "done")]
    Done,

    #[serde(rename = "interrupted")]
    Interrupted,

    #[serde(rename = "failed")]
    Failed,

    #[serde(rename = "usage_update")]
    UsageUpdate {
        #[serde(rename = "inputTokens")]
        input_tokens: u64,
        #[serde(rename = "outputTokens")]
        output_tokens: u64,
    },

    #[serde(rename = "cron_job_triggered")]
    CronJobTriggered {
        #[serde(rename = "jobId")]
        job_id: String,
        name: String,
    },

    #[serde(rename = "cron_job_completed")]
    CronJobCompleted {
        #[serde(rename = "jobId")]
        job_id: String,
        name: String,
        success: bool,
        output: String,
        #[serde(rename = "durationMs")]
        duration_ms: u64,
    },
}

File: src/shared/types/schema.ts
export interface ToolCall {
  id: string;
  name: string;
  arguments: unknown;
  rawArguments?: string;
  isPreparing?: boolean;
  result?: string;
  status: 'running' | 'completed' | 'error' | 'interrupted';
  logs?: string[];
  startedAt?: number;
  endedAt?: number;
  subSessionId?: string;
}

export interface Task {
  id: string;
  text: string;
  status: 'pending' | 'running' | 'completed' | 'error';
}

export type SelectionType = 'single_with_input' | 'multiple_with_input';

export interface AskQuestion {
  question: string;
  options: string[];
  selectionType: SelectionType;
}

export interface AskRequest {
  toolId: string;
  title: string;
  question: string;
  options: string[];
  selectionType: SelectionType;
  questions?: AskQuestion[];
}

export interface PermissionRequestEvent {
  toolId: string;
  toolName: string;
  reason: string;
}

export interface UsageSnapshot {
  inputTokens: number;
  outputTokens: number;
}

export type StreamRuntimeState =
  | 'idle'
  | 'starting'
  | 'streaming'
  | 'backoff_waiting'
  | 'retrying'
  | 'waiting_for_permission'
  | 'waiting_for_user'
  | 'interrupting'
  | 'interrupted'
  | 'completed'
  | 'failed';

export type StreamRuntimeReason =
  | 'rate_limit'
  | 'permission'
  | 'user_input'
  | 'interrupt'
  | 'completed'
  | 'error'
  | 'unknown';

export interface StreamRuntime {
  state: StreamRuntimeState;
  reason?: StreamRuntimeReason;
  message?: string;
  attempt?: number;
  retryAt?: number;
  retryInSeconds?: number;
  updatedAt: number;
}

export interface Attachment {
  id: string;
  kind: 'image' | 'file';
  name: string;
  mimeType: string;
  size: number;
  dataUrl?: string;
  previewUrl?: string;
  text?: string;
}

type EventChunkBase = {
  sessionId: string;
  eventId?: number;
  timestamp?: number;
};

export type ServerEventChunk =
  | (EventChunkBase & { type: 'runtime_status'; state: StreamRuntimeState; reason?: StreamRuntimeReason; message?: string; attempt?: number; retryAt?: number; retryInSeconds?: number })
  | (EventChunkBase & { type: 'heartbeat' })
  | (EventChunkBase & { type: 'text_delta'; content: string })
  | (EventChunkBase & { type: 'thinking_delta'; content: string })
  | (EventChunkBase & { type: 'thinking_end' })
  | (EventChunkBase & { type: 'tool_start'; toolId: string; toolName: string; args: unknown })
  | (EventChunkBase & { type: 'tool_call_delta'; toolId: string; toolName: string; argumentsText: string })
  | (EventChunkBase & { type: 'tool_output'; toolId: string; logLine: string })
  | (EventChunkBase & { type: 'tool_result'; toolId: string; result: string; isError: boolean })
  | (EventChunkBase & { type: 'tool_end'; toolId: string; exitCode: number })
  | (EventChunkBase & { type: 'ask_request'; toolId: string; title: string; question: string; options: string[]; selectionType: SelectionType; questions?: AskQuestion[] })
  | (EventChunkBase & { type: 'task_update'; tasks: Task[] })
  | (EventChunkBase & { type: 'subagent_start'; parentSessionId: string; parentToolCallId: string; subSessionId: string; title: string; message: string; startedAt: number })
  | (EventChunkBase & { type: 'subagent_end'; parentSessionId: string; parentToolCallId: string; subSessionId: string; result: string; isError: boolean })
  | (EventChunkBase & { type: 'permission_request'; toolId: string; toolName: string; reason: string })
  | (EventChunkBase & { type: 'done' })
  | (EventChunkBase & { type: 'interrupted' })
  | (EventChunkBase & { type: 'failed' })
  | (EventChunkBase & { type: 'usage_update'; inputTokens: number; outputTokens: number })
  | (EventChunkBase & { type: 'cron_job_triggered'; jobId: string; name: string })
  | (EventChunkBase & { type: 'cron_job_completed'; jobId: string; name: string; success: boolean; output: string; durationMs: number });

export interface QueuedMessage {
  id: string;
  message: Message;
  mode?: 'normal' | 'build' | 'task' | 'ask' | 'append';
  priority: 'normal' | 'urgent';
  createdAt: number;
}

export type SessionQueueState =
  | 'idle'
  | 'streaming_with_queue'
  | 'processing_queue'
  | 'interrupting';

export type MessageSegment =
  | { type: 'thinking'; content: string; isLive?: boolean; startedAt?: number; endedAt?: number }
  | { type: 'tool'; tool: ToolCall }
  | { type: 'ask'; toolId: string; title: string; question: string; options: string[]; selectionType: SelectionType; questions?: AskQuestion[]; status: 'waiting' | 'answered'; answer?: { selected: string[]; text: string }; startedAt?: number; endedAt?: number }
  | { type: 'tasks'; tasks: Task[]; startedAt?: number; endedAt?: number }
  | { type: 'retry'; state: 'backoff_waiting' | 'retrying'; reason?: StreamRuntimeReason; message: string; attempt: number; retryAt?: number; retryInSeconds?: number; updatedAt: number }
  | { type: 'text'; content: string }
  | { type: 'permission'; request: PermissionRequestEvent; status: 'waiting' | 'approved' | 'denied'; startedAt?: number; endedAt?: number };

export interface BaseMessage {
  id: string;
  attachments?: Attachment[];
  agentId?: string;
  slashCommandName?: string;
  contextPolicy?: 'default' | 'exclude';
  model?: string;
  createdAt: number;
  segments?: MessageSegment[];
  status?: 'running' | 'waiting_for_user' | 'waiting_for_permission' | 'completed';
  startedAt?: number;
  endedAt?: number;
}

export interface UserOrSystemMessage extends BaseMessage {
  role: 'user' | 'system';
  content?: string;
}

export interface AssistantMessage extends BaseMessage {
  role: 'assistant';
  content?: never;
}

export type Message = UserOrSystemMessage | AssistantMessage;

export interface Session {
  id: string;
  title: string;
  updatedAt: number;
  workspacePath?: string;
  messages: Message[];
  pinned?: boolean;
  archived?: boolean;
  hasUnreadCompleted?: boolean;
  taskDockMinimized?: boolean;
  appendDockMinimized?: boolean;
  parentId?: string;
  queuedMessages?: QueuedMessage[];
  queueState?: SessionQueueState;
  usage?: UsageSnapshot;
  tokenCount?: number;
  permissionGrants?: string[];
  subagentResult?: {
    result: string;
    isError: boolean;
    endedAt: number;
  };
  runtime?: StreamRuntime;
  error?: string | null;
}
